### PR TITLE
Move migrations to versioned tuples

### PR DESCRIPTION
**Pull request by chevdor**
WIP

## Description

Follow up of the work started in https://github.com/paritytech/polkadot/pull/7168 by @ggwpez.

This PR moves all current migrations to versioned tuples.
It allows bringing all runtimes from the SDK repo into a compliant state similar to Polkadot and Kusama.

The introduction of ~`migrations::V1_04_00`~ `migrations::V1_05_00` was not fully required yet but was introduced as placeholder to get things started for the next release after the PR is merged.

When v1.5.0 will be released, all `Unreleased` migrations can move to `migrations::V1_05_00` and free the room in the now empty `migrations::Unreleased` for the next version.

## Check

The PR introduces a new CI check for the formatting of the migrations.

## fmt

This PR introduces a few `#[rustfmt::skip]` to prevent the `Migrations` from being reformatted to stress on the importance of the migrations' ordering.

```
#[rustfmt::skip]
pub type Migrations = (
	migrations::V1_04_00,
	migrations::Unreleased
);
```

```
Checking migrations in repo: . 
./cumulus/parachains/runtimes/assets/asset-hub-polkadot/src/lib.rs:
  - ✅ V1_04_00
  - ✅ Unreleased
./cumulus/parachains/runtimes/assets/asset-hub-westend/src/lib.rs:
  - ✅ V0_09_20
  - ✅ V1_04_00
  - ✅ Unreleased
./cumulus/parachains/runtimes/assets/asset-hub-rococo/src/lib.rs:
  - ✅ V1_04_00
  - ✅ Unreleased
./cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/lib.rs:
  - ✅ V1_04_00
  - ✅ Unreleased
./cumulus/parachains/runtimes/contracts/contracts-rococo/src/lib.rs:
  - ✅ V1_04_00
  - ✅ Unreleased
./cumulus/parachains/runtimes/bridge-hubs/bridge-hub-polkadot/src/lib.rs:
  - ✅ V1_04_00
  - ✅ Unreleased
./polkadot/runtime/westend/src/lib.rs:
  - ✅ Unreleased
./cumulus/parachains/runtimes/assets/asset-hub-kusama/src/lib.rs:
  - ✅ V1_04_00
  - ✅ Unreleased
./polkadot/runtime/rococo/src/lib.rs:
  - ✅ Unreleased
./cumulus/parachains/runtimes/testing/penpal/src/lib.rs:
  - ✅ V1_04_00
  - ✅ Unreleased
./cumulus/parachains/runtimes/bridge-hubs/bridge-hub-kusama/src/lib.rs:
  - ✅ V1_04_00
  - ✅ Unreleased
```

# Checklist

- [x] My PR includes a detailed description as outlined in the "Description" section above
- [x] Fix formatting
- [x] My PR follows the [labeling requirements](CONTRIBUTING.md#Process) of this project (at minimum one label for `T`
  required)

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4085022

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4085157

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4085419

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4124411

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4129215

File: prdoc/pr_2030.prdoc [hunk 1]
```diff
@@ -0,0 +1,18 @@
+title: Move migrations to versioned tuples
+
+doc:
+  - audience: Runtime Dev
+    description: |
+      The PR [paritytech/polkadot#7168](https://github.com/paritytech/polkadot/pull/7168) introduced the grouping of
```
[the remark this answers is not in the document]
Comment (chevdor) [reply prdoc/pr_2030.prdoc]: prdoc removed

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4178580

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4179327

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4211702

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4244019

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4244871

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4259136

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4440014

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4440718

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4449308

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4452337

**Comment by chevdor** [comment]
cc @liamaharon

**Review by EgorPopelyaev — APPROVED** [review]

File: .github/workflows/check-migrations.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+          cargo install submig
+          submig --version
+
+      - name: Checkout repo
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 #v4.1.1
+
+      - name: Check migrations
+        continue-on-error: false
+        run: |
+          submig list .
```
Comment (liamaharon) [inline .github/workflows/check-migrations.yml:27]: Cool! Would be nice to have a couple sentences explaining what this does :)

File: .github/workflows/check-migrations.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+name: Check Migrations
```
Comment (liamaharon) [inline .github/workflows/check-migrations.yml:1]: To disambiguate between the try-runtime-cli checks maybe
```suggestion
name: Lint Migration Tuples
```

File: cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/lib.rs [hunk 4]
```diff
@@ -113,13 +113,39 @@ pub type SignedExtra = (
...
+	/// will be moved to a new tuple named `Vx_y_z` and added to the migration
+	/// to execute. The `Unreleased` tuple will then be cleared.
+	/// This allows keeping track of the migrations that have been applied to the
+	/// runtime *per version*.
+	// Do not remove `#[rustfmt::skip]` or the empty tuple, even if it appears
+	// to be *temporarily* not needed. Keep a single migration per line.
+	pub type Unreleased = (
+		pallet_collator_selection::migration::v1::MigrateToV1<Runtime>,
+		pallet_multisig::migrations::v1::MigrateToV1<Runtime>,
+		InitStorageVersions,
```
Comment (liamaharon) [inline cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/lib.rs:145]: are you sure these are all unreleased and were not included in the last runtime upgrade? 

same question for the other files

File: cumulus/parachains/runtimes/testing/penpal/src/lib.rs [hunk 5]
```diff
@@ -130,11 +130,34 @@ pub type SignedExtra = (
 pub type UncheckedExtrinsic =
 	generic::UncheckedExtrinsic<Address, RuntimeCall, Signature, SignedExtra>;
 
+#[rustfmt::skip]
 pub type Migrations = (
-	pallet_balances::migration::MigrateToTrackInactive<Runtime, xcm_config::CheckingAccount>,
-	pallet_collator_selection::migration::v1::MigrateToV1<Runtime>,
+	migrations::V1_05_00,
```
Comment (liamaharon) [inline cumulus/parachains/runtimes/testing/penpal/src/lib.rs:135]: Penpal is a test network we actually deploy and maintain?

**Review by liamaharon — APPROVED** [review]
Thank you for this 🙏

Looks good generally, however I think some of the migrations currently inside `Unreleased` tuples have been deployed already? 

Would be nice to get a few words on the new GA too :)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4456339

File: .github/workflows/check-migrations.yml (hunk 2 above)
Comment (chevdor) [reply .github/workflows/check-migrations.yml:27]: I could add a link to https://crates.io/crates/submig :)

Comment (chevdor) [reply .github/workflows/check-migrations.yml:27]: >  however I think some of the migrations currently inside Unreleased tuples have been deployed already?

Yes indeed, the question is which ones.....

They don't really hurt being around (for now). Once the PR is merged, we can move all of the them to the next release and flush the list. Sure the content of `V1_05_00` will then be slightly incorrect in the sense that some migrations may have been introduced earlier.

I did create older version when the info was available as comment but it was mostly not the case.

File: cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/lib.rs (hunk 4 above)
Comment (chevdor) [reply cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/lib.rs:145]: Yes, same comment than above. Without the versioning this PR brings, we cannot easily know.

File: cumulus/parachains/runtimes/testing/penpal/src/lib.rs (hunk 5 above)
Comment (chevdor) [reply cumulus/parachains/runtimes/testing/penpal/src/lib.rs:135]: It is not used much afaik but it is easier to have the versioning sorted for **all** runtimes.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4457866

File: .github/workflows/check-migrations.yml (hunk 2 above)
Comment (chevdor) [reply .github/workflows/check-migrations.yml]: I added some doc and renamed also the GHA file to make it clearer.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4458157

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4628879

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4636961

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4706691

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4706726

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4707693

**Pull request closed without merging**